### PR TITLE
Generic version of pixel_shuffle - 1d, 2d, ..., nd

### PR DESCRIPTION
**Pull request by linzwatt**
Been implementing a CNN and needed a 1D pixel shuffle, I saw PR #5051 which someone made a 3D version, saw the comments asking for a generic version, here is my attempt. First PR to a major repo, sorry if I've done something terribly wrong. Thanks.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1655,25 +1657,43 @@ def pixel_shuffle(input, upscale_factor):
         upscale_factor (int): factor to increase spatial resolution by
 
     Examples::
+        # 1D example
+        >>> ps = nn.PixelShuffle(2)
+        >>> input = autograd.Variable(torch.Tensor(1, 4, 8))
```
Comment (SsnL) [inline torch/nn/functional.py]: lets remove Variable wrappers in example code

**Comment by Stonesjtu** [comment]
You should add the corresponding Modules at the same time, I think.

**Comment by lindsaywatt** [comment]
@Stonesjtu do you mean like `PixelShuffle1D(nn.Module)`, `PixelShuffle2D(nn.Module)`, `PixelShuffle3D(nn.Module)` etc? I think that would defeat the purpose of having this generic nD version

**Comment by Stonesjtu** [comment]
The generic version `PixelShuffle(nn.Module)` should be fine.

**Comment by lindsaywatt** [comment]
No code needs to change in the PixelShuffle Module for this to work, but I will make another commit with updated doc string and example for the Module

**Comment by ezyang** [comment]
Thanks @lindsaywatt, looking forward for the commit.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1655,25 +1656,40 @@ def pixel_shuffle(input, upscale_factor):
...
+    output_size = [dim * upscale_factor for dim in input_size[2:]]
 
     input_view = input.contiguous().view(
-        batch_size, channels, upscale_factor, upscale_factor,
-        in_height, in_width)
+        input_size[0], input_size[1],
+        *(([upscale_factor] * dimensionality) + input_size[2:])
+    )
+
+    indicies = list(range(2, 2 + 2 * dimensionality))
```
Comment (ezyang) [inline torch/nn/functional.py]: indices

**Review by ezyang — CHANGES_REQUESTED** [review]
This looks good but we need a test exercising the higher dimensionality case.

**Comment by ezyang** [comment]
(OK to merge once the test is ready, ping me if I forget to look again.)

**Comment by zou3519** [comment]
@lindsaywatt can you add a test according to what @ezyang mentioned?

**Comment by ezyang** [comment]
Reping @lindsaywatt

**Comment by weiyangfb** [comment]
@lindsaywatt any update on this PR?

File: torch/nn/functional.py (hunk 1 above)
Comment (myhub) [reply torch/nn/functional.py]: I think pixel_shuffle 1D is not what we want
<pre>x = torch.arange(8).view(1, 2, 4)
print(x)
print(x.shape)
y = pixel_shuffle(x, 2)
print(y.shape)
print(y)
sys.exit()<pre>
the output is tensor([[[0, 1, 2, 3, 4, 5, 6, 7]]])
but we expect [0, 4, 1, 5, 2, 6, 3, 7]
pixel_shuffle 2d is right

Comment (myhub) [reply torch/nn/functional.py]: As you can see, pixel_shuffle 1D  just do like torch.cat(),  not pixel_shuffle
@SsnL

**Comment by davidaknowles** [comment]
In case anyone ends up here and wants the 1d version: https://gist.github.com/davidaknowles/6e95a643adaf3960d1648a6b369e9d0b

**Comment by LukasMosser** [comment]
Great that someone picked this up! Thanks!

**Review by feifeife — APPROVED** [review]
